src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -449,3 +556,153 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
+	FuncExpr *funcExpr = NULL;
+
+	if (list_length(targetList) > 1 || list_length(joinTree->fromlist) > 1)
+	{
+		/* multiple target list or from list items */
+		return;
+	}
+
+	/*
+	 * Look for a function in the SELECT clause.
+	 */
+	if (list_length(targetList) == 1)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) linitial(targetList);
+		if (IsA(targetEntry->expr, FuncExpr))
+		{
+			funcExpr = (FuncExpr *) targetEntry->expr;
+		}
+	}
+
+	/*
+	 * Look for a function in the FROM clause.
+	 */
+	if (list_length(joinTree->fromlist) == 1)
+	{
+		RangeTblRef *reference = linitial(joinTree->fromlist);
+		if (IsA(reference, RangeTblRef))
+		{
+			RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
+											  planContext->query->rtable);
+
+			if (rtentry->rtekind == RTE_FUNCTION && list_length(rtentry->functions) == 1)
+			{
+				RangeTblFunction *rtfunc = (RangeTblFunction *) linitial(
+					rtentry->functions);
+				if (IsA(rtfunc->funcexpr, FuncExpr))
+				{
+					if (funcExpr)
+					{
+						/* Function in both the SELECT and FROM list */
+						return;
+					}
+					else
+					{
+						funcExpr = (FuncExpr *) rtfunc->funcexpr;
+					}
+				}
+			}
+		}
+	}
+
+	if (!funcExpr)
+	{
+		/* No function in either SELECT or FROM list */
+		return;
+	}
+
+	DistObjectCacheEntry *procedure = LookupDistObjectCacheEntry(ProcedureRelationId,
+																 funcExpr->funcid, 0);
+	Assert(procedure != NULL);
+
+	if (procedure->isDistributed && procedure->forcePushdown)
+	{
+		/*
+		 * On the coordinator PartiallyEvaluateExpression() descends into an
+		 * expression tree to evaluate expressions that can be resolved to a
+		 * constant. Expressions containing a Var are skipped, since the value
+		 * of the Var is not known on the coordinator.
+		 */
+		Node *partitionValueNode = (Node *) list_nth(funcExpr->args,
+													 procedure->distributionArgIndex);
+		Assert(partitionValueNode);
+		partitionValueNode = strip_implicit_coercions(partitionValueNode);
+
+		if (IsA(partitionValueNode, Param))
+		{
+			Param *partitionParam = (Param *) partitionValueNode;
+
+			if (partitionParam->paramkind == PARAM_EXTERN)
+			{
+				/* Don't log a message, we should end up here again without a parameter */
+				DissuadePlannerFromUsingPlan(planContext->plan);
+				return;
+			}
+		}
+
+		if (IsA(partitionValueNode, Const))
+		{
+			elog(DEBUG1, "Pushdown argument: %s",
+				 pretty_format_node_dump(nodeToString(partitionValueNode)));
+			EnableInForceDelegatedFuncExecution((Const *) partitionValueNode);
+		}
+		else
+		{
+			/*
+			 * Var node e.g. select fn(col) from table where col=150;
+			 * Param(PARAM_EXEC) node e.g. SELECT fn((SELECT col from test_nested where col=val))
+			 */
+			elog(NOTICE, "Non-constant Pushdown argument: ");
+			ereport(ERROR, (errmsg(
+								"Non-constant arguments for forcePushdown functions not supported")));
+		}
+	}
+	else
+	{
+		/* not a delegated function call within a 2PC */
+		return;
```

Comment:
would prefer early return

Reply:
Earliest we can do is once we get the **_FuncExpr_**. Moved it.